File: RMG.Core/Composition/TrackEventStatePattern.fs
namespace RMG.Core.Composition

open RMG.Core

[<Sealed>]
type TrackEventStatePattern
    private
    (
        duration: Duration,
        trackNumberMap: Map<TrackNumber option, TrackNumber option> option,
        trackPatternTimelineMap: Map<TrackNumber option, EventStatePattern WithEventState Timeline>,
        trackPatternMapTimeline: TrackEventStatePattern WithEventState Timeline,
        flatTimelineMap: TrackEventStateTimelineMap
    ) =
    member this.Duration = duration
    member this.TrackNumberMap = trackNumberMap
    member this.TrackPatternTimelineMap = trackPatternTimelineMap
    member this.TrackPatternMapTimeline = trackPatternMapTimeline
    member this.FlatTimelineMap = flatTimelineMap

    static member internal ofTimelineMapsUnsafe
        (duration: Duration)
        (trackNumberMap: Map<TrackNumber option, TrackNumber option> option)
        (trackPatternTimelineMap: Map<TrackNumber option, EventStatePattern WithEventState Timeline>)
        (trackPatternMapTimeline: TrackEventStatePattern WithEventState Timeline)
        (flatTimelineMap: TrackEventStateTimelineMap)
        : TrackEventStatePattern =
        TrackEventStatePattern(duration, trackNumberMap, trackPatternTimelineMap, trackPatternMapTimeline, flatTimelineMap)

module TrackEventStatePattern =
    let ofTimelineMaps
        (duration: Duration)
        (trackNumberMap: Map<TrackNumber option, TrackNumber option> option)
        (trackPatternTimelineMap: Map<TrackNumber option, EventStatePattern WithEventState Timeline>)
        (trackPatternMapTimeline: TrackEventStatePattern WithEventState Timeline)
        : TrackEventStatePattern
        =

        let mapTrackNumbers (inputTimelineMap: TrackEventStateTimelineMap) : TrackEventStateTimelineMap =
            match trackNumberMap with
            | Some map -> inputTimelineMap |> TrackEventStateTimelineMap.mapTrackNumbers map
            | _ -> inputTimelineMap

        let flatTimelineMap =
            seq {
                yield
                    trackPatternTimelineMap
                    |> Map.map
                        (fun _ patternTimeline ->
                            patternTimeline
                            |> Timeline.map (fun struct (eventState, pattern) -> struct (eventState, pattern.FlatTimeline))
                            |> EventStateTimelineMap.concatCombined duration)
                    |> Timeline.itemOfSingle

                yield!
                    trackPatternMapTimeline
                    |> Timeline.map
                        (fun struct (eventState, pattern) ->
                            pattern.FlatTimelineMap
                            |> mapTrackNumbers
                            |> TrackEventStateTimelineMap.shiftState eventState)
            }
            |> Timeline.ofSeq
            |> TrackEventStateTimelineMap.concat duration

        TrackEventStatePattern.ofTimelineMapsUnsafe duration trackNumberMap trackPatternTimelineMap trackPatternMapTimeline flatTimelineMap

File: Rmg.Core/Composition/TrackEventStatePattern.cs
using Rmg.Core.Events;

namespace Rmg.Core.Composition;

public sealed class TrackEventStatePattern<T>
    where T : notnull
{
    public static TrackEventStatePattern<T> Create(
        double duration,
        IEnumerable<KeyValuePair<int, int>> trackIndexMap,
        TrackEventStateTimelineMap<WithStateMap<EventStatePattern<T>>> trackPatternTimelineMap,
        EventTimeline<WithStateMap<TrackEventStatePattern<T>>> trackPatternMapTimeline
    )
    {
        ArgumentOutOfRangeException.ThrowIfNegative(duration);

        var trackIndexDictionary = trackIndexMap.ToDictionary();

        var unwrappedTrackPatternTimelineMap = trackPatternTimelineMap
            .MapTrackTimelines(eventStateTimelineMap => eventStateTimelineMap
                .MapEventTimeline(eventTimeline => eventTimeline.MapValues(x => x.Value.MergeStateMap(x.StateMap)))
                .Unwrap()
            );
        var unwrappedTrackPatternMapTimeline = trackPatternMapTimeline
            .MapValues(x => x.Value.MergeStateMap(x.StateMap))
            .Unwrap();
        var mergedStateTimelineMap =
            TrackEventStateTimelineMap.Merge([unwrappedTrackPatternTimelineMap, unwrappedTrackPatternMapTimeline])
                .MapTrackIndexes(trackIndexDictionary);

        return new TrackEventStatePattern<T>(
            duration,
            trackIndexDictionary,
            trackPatternTimelineMap,
            trackPatternMapTimeline,
            mergedStateTimelineMap
        );
    }

    public double Duration { get; }

    public IReadOnlyDictionary<int, int> TrackIndexMap { get; }

    public TrackEventStateTimelineMap<WithStateMap<EventStatePattern<T>>> TrackPatternTimelineMap { get; }

    public EventTimeline<WithStateMap<TrackEventStatePattern<T>>> TrackPatternMapTimeline { get; }

    public TrackEventStateTimelineMap<T> MergedTimelineMap { get; }

    private TrackEventStatePattern(
        double duration,
        IReadOnlyDictionary<int, int> trackIndexMap,
        TrackEventStateTimelineMap<WithStateMap<EventStatePattern<T>>> trackPatternTimelineMap,
        EventTimeline<WithStateMap<TrackEventStatePattern<T>>> trackPatternMapTimeline,
        TrackEventStateTimelineMap<T> mergedTimelineMap
    )
    {
        ArgumentOutOfRangeException.ThrowIfNegative(duration);

        Duration = duration;
        TrackIndexMap = trackIndexMap;
        TrackPatternTimelineMap = trackPatternTimelineMap;
        TrackPatternMapTimeline = trackPatternMapTimeline;
        MergedTimelineMap = mergedTimelineMap;
    }

    public TrackEventStateTimelineMap<T> MergeStateMap(StateMap stateMap)
    {
        return MergedTimelineMap.MergeStateMap(stateMap);
    }
}

public static class TrackEventStatePattern
{
    public static TrackEventStatePattern<T> Create<T>(
        double duration,
        IEnumerable<KeyValuePair<int, int>> trackIndexMap,
        TrackEventStateTimelineMap<WithStateMap<EventStatePattern<T>>> trackPatternTimelineMap,
        EventTimeline<WithStateMap<TrackEventStatePattern<T>>> trackPatternMapTimeline
    ) where T : notnull
    {
        return TrackEventStatePattern<T>.Create(
            duration,
            trackIndexMap,
            trackPatternTimelineMap,
            trackPatternMapTimeline
        );
    }
}